core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java

```diff
@@ -0,0 +1,216 @@
+    // Can't reset if we're not initialized
+    if (!sInit) {
+      return;
+    }
+
+    // Reset state
+    sInit = false;
+    FACTORIES.clear();
+
+    // Reinitialise
+    init();
+  }
+
+  /**
+   * Adds a new factory
+   * <p>
+   * Factories are added to the start of the factories list so they can override the existing
+   * automatically discovered factories. Generally if you use the ServiceLoader mechanism properly
+   * it should be unnecessary to call this, however since ServiceLoader discovery order may be
+   * susceptible to class loader behavioural differences there may be rare cases when you need to
+   * manually register the desired factory.
+   * <p>
+   * 
+   * @param factory
+   *          Factory to add
+   */
+  public static void add(UnderFileSystemFactory factory) {
+    if (factory == null) {
+      return;
+    }
+
+    LOG.debug("Registered Under File System Factory implementation {} - {}", factory.getClass(),
+        factory.toString());
+    FACTORIES.add(0, factory);
+  }
+
+  /**
+   * Removes an existing factory
+   * 
+   * @param factory
+   *          Factory to remove
+   */
+  public static void remove(UnderFileSystemFactory factory) {
+    if (factory == null) {
+      return;
+    }
+
+    LOG.debug("Unregistered Under File System Factory implementation {} - {}", factory.getClass(),
+        factory.toString());
+    FACTORIES.remove(factory);
+  }
+
+  /**
+   * Returns a read-only view of the available factories
+   * 
+   * @return Read-only view of the available factories
+   */
+  public static List<UnderFileSystemFactory> available() {
+    return Collections.unmodifiableList(FACTORIES);
+  }
+
+  /**
+   * Tries to find a Under File System factory that supports the given path
+   * 
+   * @param path
+   *          Path
+   * @return Factory if available, null otherwise
+   */
+  public static UnderFileSystemFactory find(String path) {
+    Preconditions.checkArgument(path != null, "path may not be null");
+
+    for (UnderFileSystemFactory factory : FACTORIES) {
+      if (factory.supportsPath(path)) {
+        LOG.debug("Selected Under File System Factory implementation {} for path {}",
+            factory.getClass(), path);
+        return factory;
+      }
+    }
+
+    LOG.warn("No Under File System Factory implementation supports the path {}", path);
+    return null;
+  }
+
+  /**
+   * Creates a client that can talk to the under file system
+   * 
+   * @param path
+   *          Path
+   * @param conf
+   *          Optional configuration object, may be null
+   * @return Client for the under file system
+   * @throws IllegalArgumentException
+   *           Thrown if there is no under file system for the given path
+   */
+  public static UnderFileSystem create(String path, Object conf) {
+    // Try to obtain the appropriate factory
+    UnderFileSystemFactory factory = find(path);
+    if (factory == null) {
+      throw new IllegalArgumentException(String.format(
```

Comment:
Sure, this is just my stylistic preference but I'm happy to go with whatever is preferred